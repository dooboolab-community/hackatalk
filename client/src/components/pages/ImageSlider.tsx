import {Animated, Dimensions, Image, StatusBar} from 'react-native';
import PinchZoom, {PinchZoomRef} from '../uis/PinchZoom';
import {
  RootStackNavigationProps,
  RootStackParamList,
} from '../navigations/RootStackNavigator';
import {RouteProp, useNavigation} from '@react-navigation/native';

import React from 'react';
import styled from 'styled-components/native';
import {useSafeAreaInsets} from 'react-native-safe-area-context';

const Container = styled.View`
  flex: 1;
  background-color: #000;
  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

const ImageSliderContainer = styled.View`
  flex: 1;
  justify-content: center;
  max-width: ${(): number => Dimensions.get('screen').width}px;
`;

interface Props {
  navigation: RootStackNavigationProps<'ImageSlider'>;
  route: RouteProp<RootStackParamList, 'ImageSlider'>;
}

function ImageSlider({
  route: {
    params: {images, initialIndex = 0},
  },
}: Props): React.ReactElement {
  const [currentIndex, setCurrentIndex] = React.useState<number>(initialIndex);
  const insets = useSafeAreaInsets();
  const imageContainerWidth = Dimensions.get('screen').width;

  const imageContainerHeight =
    Dimensions.get('screen').height - insets.top - insets.bottom;

  const imageStyle = {
    width: imageContainerWidth,
    bottom: 0,
    height: imageContainerHeight,
  };

  const imageGap = 10;

  const animValues = React.useRef({
    scale: 1,
    x: 0,
    y: 0,
    nextTranslateX: 0,
    prevTranslateX: 0,
  }).current;

  const pinchZoom = React.useRef<PinchZoomRef>(null);
  const nextImageTranslateX = React.useRef(new Animated.Value(0)).current;
  const prevImageTranslateX = React.useRef(new Animated.Value(0)).current;
  const navigation = useNavigation();

  const translateOtherImages = React.useCallback(() => {
    animValues.nextTranslateX =
      ((animValues.scale - 1) * imageContainerWidth) / 2 + animValues.x;

    animValues.prevTranslateX =
      ((1 - animValues.scale) * imageContainerWidth) / 2 + animValues.x;

    nextImageTranslateX.setValue(animValues.nextTranslateX);
    prevImageTranslateX.setValue(animValues.prevTranslateX);
  }, [
    imageContainerWidth,
    animValues,
    nextImageTranslateX,
    prevImageTranslateX,
  ]);

  const onRelease = React.useCallback(() => {
    const moveNext = animValues.nextTranslateX < -imageContainerWidth / 2;
    const movePrev = animValues.prevTranslateX > imageContainerWidth / 2;
    const targetTranslate = pinchZoom.current?.animatedValue.translate;

    if (moveNext && currentIndex < images.length - 1 && targetTranslate)
      Animated.timing(targetTranslate, {
        toValue: {
          x: (-(animValues.scale + 1) / 2) * imageContainerWidth,
          y: animValues.y,
        },
        useNativeDriver: true,
        duration: 300,
      }).start(() => {
        setCurrentIndex(currentIndex + 1);
      });
    else if (movePrev && currentIndex > 0 && targetTranslate)
      Animated.timing(targetTranslate, {
        toValue: {
          x: ((animValues.scale + 1) / 2) * imageContainerWidth,
          y: animValues.y,
        },
        useNativeDriver: true,
        duration: 300,
      }).start(() => {
        nextImageTranslateX.setValue(0);
        prevImageTranslateX.setValue(0);
        setCurrentIndex(currentIndex - 1);
      });
    else if (animValues.nextTranslateX < 0 && targetTranslate)
      Animated.timing(targetTranslate, {
        toValue: {
          x: ((1 - animValues.scale) * imageContainerWidth) / 2,
          y: animValues.y,
        },
        useNativeDriver: true,
        duration: 300,
      }).start(() => {
        pinchZoom.current?.setValues({
          translate: {
            x: ((1 - animValues.scale) * imageContainerWidth) / 2,
            y: animValues.y,
          },
        });
      });
    else if (animValues.prevTranslateX > 0 && targetTranslate)
      Animated.timing(targetTranslate, {
        toValue: {
          x: ((animValues.scale - 1) * imageContainerWidth) / 2,
          y: animValues.y,
        },
        useNativeDriver: true,
        duration: 300,
      }).start(() => {
        pinchZoom.current?.setValues({
          translate: {
            x: ((animValues.scale - 1) * imageContainerWidth) / 2,
            y: animValues.y,
          },
        });
      });
  }, [
    currentIndex,
    imageContainerWidth,
    animValues.nextTranslateX,
    animValues.prevTranslateX,
    animValues.scale,
    animValues.y,
    images.length,
    nextImageTranslateX,
    prevImageTranslateX,
  ]);

  React.useEffect(() => {
    navigation.setOptions({
      headerBackTitle: images[currentIndex]?.sender || '',
    });

    pinchZoom.current?.setValues({scale: 1, translate: {x: 0, y: 0}});
    nextImageTranslateX.setValue(0);
    prevImageTranslateX.setValue(0);
  }, [
    currentIndex,
    images,
    navigation,
    nextImageTranslateX,
    prevImageTranslateX,
  ]);

  return (
    <Container>
      <StatusBar barStyle="light-content" backgroundColor="#6a51ae" />
      <ImageSliderContainer>
        {currentIndex > 0 ? (
          <PinchZoom
            key={`${currentIndex - 1}`}
            style={{
              position: 'absolute',
              width: imageContainerWidth,
              left: -imageContainerWidth - imageGap,
              top: 0,
              bottom: 0,
              justifyContent: 'center',
              // @ts-ignore
              transform: [{translateX: prevImageTranslateX}],
            }}>
            <Image
              source={{uri: images[currentIndex - 1]?.uri as string}}
              style={imageStyle}
              resizeMode="contain"
            />
          </PinchZoom>
        ) : null}
        <PinchZoom
          key={`${currentIndex}`}
          ref={pinchZoom}
          onScaleChanged={(value): void => {
            animValues.scale = value;
            translateOtherImages();
          }}
          onTranslateChanged={(value): void => {
            animValues.x = value.x;
            animValues.y = value.y;
            translateOtherImages();
          }}
          onRelease={onRelease}
          allowEmpty={{x: true}}
          fixOverflowAfterRelease={false}
          style={{
            width: imageContainerWidth,
            justifyContent: 'center',
          }}>
          <Image
            source={{uri: images[currentIndex]?.uri as string}}
            style={imageStyle}
            resizeMode="contain"
          />
        </PinchZoom>
        {currentIndex < images.length - 1 ? (
          <PinchZoom
            key={`${currentIndex + 1}`}
            style={{
              width: imageContainerWidth,
              position: 'absolute',
              top: 0,
              bottom: 0,
              justifyContent: 'center',
              left: imageContainerWidth + imageGap,
              // @ts-ignore
              transform: [{translateX: nextImageTranslateX}],
            }}>
            <Image
              source={{uri: images[currentIndex + 1]?.uri as string}}
              style={imageStyle}
              resizeMode="contain"
            />
          </PinchZoom>
        ) : null}
      </ImageSliderContainer>
    </Container>
  );
}

export default ImageSlider;
