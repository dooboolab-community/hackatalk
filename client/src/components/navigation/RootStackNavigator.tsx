import {
  StackNavigationProp,
  createStackNavigator,
} from '@react-navigation/stack';

import AuthStack from './AuthStackNavigator';
import ImageSlider from '../screen/ImageSlider';
import MainStack from './MainStackNavigator';
import {NavigationContainer} from '@react-navigation/native';
import NotFound from '../screen/NotFound';
import {Platform} from 'react-native';
import React from 'react';
import {SafeAreaProvider} from 'react-native-safe-area-context';
import WebView from '../screen/WebView';
import {useAuthContext} from '../../providers/AuthProvider';
import {useThemeContext} from '@dooboo-ui/theme';

export type RootStackParamList = {
  default: undefined;
  AuthStack: undefined;
  MainStack: undefined;
  WebView: {
    uri: string;
  };
  ImageSlider: {
    images: {uri: string | null; sender?: string | null}[];
    initialIndex?: number;
  };
  NotFound: undefined;
};

export type RootStackNavigationProps<
  T extends keyof RootStackParamList = 'default'
> = StackNavigationProp<RootStackParamList, T>;

const Stack = createStackNavigator<RootStackParamList>();

function RootNavigator(): React.ReactElement {
  const {theme} = useThemeContext();

  const {
    state: {user},
  } = useAuthContext();

  return (
    <SafeAreaProvider>
      <NavigationContainer
        theme={{
          colors: {
            background: theme.background,
            border: theme.border,
            card: theme.itemBackground,
            primary: theme.primary,
            notification: theme.tintColor,
            text: theme.text,
          },
          dark: true,
        }}>
        <Stack.Navigator
          screenOptions={{
            headerStyle: {
              backgroundColor: theme.background,
            },
            headerTitleStyle: {color: theme.fontColor},
            headerTintColor: theme.tintColor,
            headerShown: false,
          }}>
          {!user || !user.verified ? (
            <Stack.Screen
              name="AuthStack"
              component={AuthStack}
              options={{
                gestureDirection: Platform.select({
                  ios: !user ? 'horizontal-inverted' : 'horizontal',
                  default: !user ? 'vertical-inverted' : 'vertical',
                }),
              }}
            />
          ) : (
            <Stack.Screen
              name="MainStack"
              component={MainStack}
              options={{
                gestureDirection: Platform.select({
                  ios: !user ? 'horizontal-inverted' : 'horizontal',
                  default: !user ? 'vertical-inverted' : 'vertical',
                }),
              }}
            />
          )}
          <Stack.Screen
            name="WebView"
            component={WebView}
            options={{
              headerShown: true,
              headerTitle: '',
              headerBackTitle: '',
            }}
          />
          <Stack.Screen
            name="ImageSlider"
            component={ImageSlider}
            options={({
              route: {
                params: {images, initialIndex = 0},
              },
            }) => ({
              headerShown: true,
              headerBackTitle: images[initialIndex]?.sender || '',
              headerTitle: '',
              headerTransparent: true,
            })}
          />
          <Stack.Screen name="NotFound" component={NotFound} />
        </Stack.Navigator>
      </NavigationContainer>
    </SafeAreaProvider>
  );
}

export default RootNavigator;
