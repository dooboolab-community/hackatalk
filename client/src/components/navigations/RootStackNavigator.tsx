import * as SplashScreen from 'expo-splash-screen';

import AuthStack, {AuthStackParamList} from './AuthStackNavigator';
import {LoadingIndicator, useTheme} from 'dooboo-ui';
import MainStack, {MainStackParamList} from './MainStackNavigator';
import {
  NavigationContainer,
  NavigatorScreenParams,
} from '@react-navigation/native';
import {Platform, View} from 'react-native';
import {PreloadedQuery, usePreloadedQuery} from 'react-relay';
import React, {FC, Suspense, useEffect} from 'react';
import {
  StackNavigationProp,
  createStackNavigator,
} from '@react-navigation/stack';
import {meQuery, useAuthContext} from '../../providers/AuthProvider';

import AsyncStorage from '@react-native-async-storage/async-storage';
import {AuthProviderMeQuery} from '../../__generated__/AuthProviderMeQuery.graphql';
import ImageSlider from '../pages/ImageSlider';
import NotFound from '../pages/NotFound';
import {SafeAreaProvider} from 'react-native-safe-area-context';
import WebView from '../pages/WebView';

export type RootStackParamList = {
  default: undefined;
  AuthStack: NavigatorScreenParams<AuthStackParamList>;
  MainStack: NavigatorScreenParams<MainStackParamList>;
  WebView: {
    uri: string;
  };
  ImageSlider: {
    images: {uri: string | null; sender?: string | null}[];
    initialIndex?: number;
  };
  NotFound: undefined;
};

export type RootStackNavigationProps<
  T extends keyof RootStackParamList = 'default',
> = StackNavigationProp<RootStackParamList, T>;

const Stack = createStackNavigator<RootStackParamList>();

type Props = {
  queryReference: PreloadedQuery<AuthProviderMeQuery, Record<string, unknown>>;
};

function RootNavigator({queryReference}: Props): React.ReactElement {
  const {theme} = useTheme();

  const linking = {
    prefixes: ['https://hackatalk.dev', 'hackatalk://'],
    enabled: true,
  };

  const {me} = usePreloadedQuery<AuthProviderMeQuery>(meQuery, queryReference);

  const {user, setUser} = useAuthContext();

  const hideSplash = async (): Promise<void> => {
    try {
      await SplashScreen.hideAsync();
    } catch (err) {
      // err
    }
  };

  useEffect(() => {
    if (!me) {
      AsyncStorage.removeItem('token');
      hideSplash();
      setUser(null);

      return;
    }

    hideSplash();

    setUser(me);
  }, [me, setUser]);

  return (
    <SafeAreaProvider>
      <NavigationContainer
        linking={Platform.select({
          web: linking,
        })}
        theme={{
          colors: {
            background: theme.background,
            border: theme.disabled,
            card: theme.itemBackground,
            primary: theme.primary,
            notification: theme.tintColor,
            text: theme.text,
          },
          dark: true,
        }}>
        <Stack.Navigator
          initialRouteName="AuthStack"
          screenOptions={{
            headerStyle: {
              backgroundColor: theme.background,
            },
            headerTitleStyle: {color: theme.text},
            headerTintColor: theme.tintColor,
            headerShown: false,
          }}>
          {!user || !user.verified ? (
            <Stack.Screen
              name="AuthStack"
              component={AuthStack}
              options={{
                gestureDirection: Platform.select({
                  ios: !user ? 'horizontal-inverted' : 'horizontal',
                  default: !user ? 'vertical-inverted' : 'vertical',
                }),
              }}
            />
          ) : (
            <Stack.Screen
              name="MainStack"
              component={MainStack}
              options={{
                gestureDirection: Platform.select({
                  ios: !user ? 'horizontal-inverted' : 'horizontal',
                  default: !user ? 'vertical-inverted' : 'vertical',
                }),
              }}
            />
          )}
          <Stack.Screen
            name="WebView"
            component={WebView}
            options={{
              headerShown: true,
              headerTitle: '',
              headerBackTitle: '',
            }}
          />
          <Stack.Screen
            name="ImageSlider"
            component={ImageSlider}
            options={({
              route: {
                params: {images, initialIndex = 0},
              },
            }) => ({
              headerShown: true,
              headerBackTitle: images[initialIndex]?.sender || '',
              headerTitle: '',
              headerTransparent: true,
            })}
          />
          <Stack.Screen name="NotFound" component={NotFound} />
        </Stack.Navigator>
      </NavigationContainer>
    </SafeAreaProvider>
  );
}

function AuthNavigatorOnly(): React.ReactElement {
  const {theme} = useTheme();

  return (
    <SafeAreaProvider>
      <NavigationContainer
        theme={{
          colors: {
            background: theme.background,
            border: theme.dialog,
            card: theme.paper,
            primary: theme.primary,
            notification: theme.primary,
            text: theme.primaryText,
          },
          dark: true,
        }}>
        <AuthStack />
      </NavigationContainer>
    </SafeAreaProvider>
  );
}

const RootNavigatorWrapper: FC = () => {
  const {theme} = useTheme();
  const {user, loadMeQuery, meQueryReference} = useAuthContext();

  useEffect(() => {
    if (!user) loadMeQuery({});
  }, [loadMeQuery, user]);

  return (
    <Suspense
      fallback={
        <View style={{flex: 1, backgroundColor: theme.background}}>
          <LoadingIndicator />
        </View>
      }>
      {meQueryReference ? (
        <RootNavigator queryReference={meQueryReference} />
      ) : (
        <AuthNavigatorOnly />
      )}
    </Suspense>
  );
};

export default RootNavigatorWrapper;
