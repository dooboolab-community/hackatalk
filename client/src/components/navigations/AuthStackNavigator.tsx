import React, {ReactElement} from 'react';
import {
  StackNavigationProp,
  createStackNavigator,
} from '@react-navigation/stack';

import {CompositeNavigationProp} from '@react-navigation/native';
import FindPwContainer from '../pages/FindPwContainer';
import LicenseAgreementContainer from '../pages/LicenseAgreementContainer';
import {RootStackNavigationProps} from './RootStackNavigator';
import SignInContainer from '../pages/SignInContainer';
import SignUp from '../pages/SignUp';
import VerifyEmail from '../pages/VerifyEmail';
import {getString} from '../../../STRINGS';
import {useTheme} from 'dooboo-ui';

export type AuthStackParamList = {
  default: undefined;
  SignIn: undefined;
  SignUp: undefined;
  FindPw: undefined;
  VerifyEmail: {
    email: string;
  };
  LicenseAgreement: undefined;
};

type NavigationProps<
  T extends keyof AuthStackParamList = 'default'
> = StackNavigationProp<AuthStackParamList, T>;

export type AuthStackNavigationProps<
  T extends keyof AuthStackParamList = 'default'
> = CompositeNavigationProp<
  NavigationProps<T>,
  RootStackNavigationProps<'AuthStack'>
>;

const Stack = createStackNavigator<AuthStackParamList>();

function AuthNavigator(): ReactElement {
  const {theme} = useTheme();

  return (
    <Stack.Navigator
      initialRouteName="SignIn"
      screenOptions={{
        headerStyle: {
          backgroundColor: theme.background,
          borderBottomWidth: 0,
          elevation: 0,
        },
        headerTintColor: theme.text,
      }}>
      <Stack.Screen
        name="SignIn"
        component={SignInContainer}
        options={{
          headerShown: false,
        }}
      />
      <Stack.Screen
        name="SignUp"
        component={SignUp}
        options={{
          title: getString('SIGN_UP'),
        }}
      />
      <Stack.Screen
        name="FindPw"
        component={FindPwContainer}
        options={{
          title: getString('FIND_PW'),
        }}
      />
      <Stack.Screen
        name="VerifyEmail"
        component={VerifyEmail}
        options={{
          title: getString('VERIFY_EMAIL'),
        }}
      />
      <Stack.Screen
        name="LicenseAgreement"
        component={LicenseAgreementContainer}
        options={{
          title: getString('LICENSE_AGREEMENT'),
        }}
      />
    </Stack.Navigator>
  );
}

export default AuthNavigator;
