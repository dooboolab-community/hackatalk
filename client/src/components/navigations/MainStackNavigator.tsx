import * as Notifications from 'expo-notifications';

import {Channel, User} from '../../types/graphql';
import {CompositeNavigationProp, useNavigation} from '@react-navigation/native';
import {Image, TouchableOpacity, View} from 'react-native';
import React, {ReactElement, useEffect, useRef} from 'react';
import {
  StackNavigationOptions,
  StackNavigationProp,
  createStackNavigator,
} from '@react-navigation/stack';
import TabNavigator, {MainTabNavigationOptions} from './MainTabNavigator';
import {fetchQuery, graphql, useRelayEnvironment} from 'react-relay/hooks';

import BlockedUser from '../pages/BlockedUser';
import ChangePw from '../pages/ChangePw';
import ChannelCreate from '../pages/ChannelCreate';
import {DefaultTheme} from 'styled-components';
import {IC_SETTING_W} from '../../utils/Icons';
import type {MainStackNavigatorChannelQuery} from '../../__generated__/MainStackNavigatorChannelQuery.graphql';
import Message from '../pages/Message';
import ProfileModal from '../templates/ProfileModal';
import {ProfileModalProvider} from '../../providers/ProfileModalProvider';
import ProfileUpdate from '../pages/ProfileUpdate';
import Report from '../pages/Report';
import {RootStackNavigationProps} from './RootStackNavigator';
import SearchUser from '../pages/SearchUser';
import Settings from '../pages/Settings';
import StatusBar from '../UI/atoms/StatusBar';
import {getString} from '../../../STRINGS';
// import useAppState from '../../hooks/useAppState';
import {useThemeContext} from '@dooboo-ui/theme';

export type MainStackParamList = {
  MainTab: undefined;
  ProfileUpdate: undefined;
  SearchUser: undefined;
  BlockedUser: undefined;
  Report: {
    name: string;
    userId: string;
  };
  Message: {
    channel: Channel;
    users?: User[];
  };
  Settings: undefined;
  ChangePw: undefined;
  ChannelCreate: undefined;
  PinchZoomViewPager: undefined;
};

type NavigationProps<T extends keyof MainStackParamList> = StackNavigationProp<
  MainStackParamList,
  T
>;

export type MainStackNavigationProps<
  T extends keyof MainStackParamList
> = CompositeNavigationProp<
  NavigationProps<T>,
  RootStackNavigationProps<'MainStack'>
>;

const Stack = createStackNavigator<MainStackParamList>();

function getSimpleHeader(
  title: string,
  theme: DefaultTheme,
): StackNavigationOptions {
  return {
    headerTitle: title,
    headerTintColor: theme.headerFont,
    headerStyle: {
      backgroundColor: theme.primary,
      borderBottomWidth: 0,
      elevation: 0,
      shadowColor: 'transparent',
    },
  };
}

const channelQuery = graphql`
  query MainStackNavigatorChannelQuery($channelId: String!) {
    channel(channelId: $channelId) {
      id
      channelType
      name
      memberships(excludeMe: true) {
        user {
          name
          nickname
          thumbURL
          photoURL
        }
      }
    }
  }
`;

function MainStackNavigator(): ReactElement {
  const {theme} = useThemeContext();
  // const currentAppState = useAppState();
  const navigation = useNavigation<MainStackNavigationProps<'MainTab'>>();
  const environment = useRelayEnvironment();

  const latestNavigation = useRef(navigation);
  const latestEnvironment = useRef(environment);

  useEffect(() => {
    const subscription = Notifications.addNotificationResponseReceivedListener(
      async (response) => {
        Notifications.setBadgeCountAsync(0);

        const channelId = JSON.parse(
          response.notification.request.content.data.data as string,
        ).channelId;

        fetchQuery<MainStackNavigatorChannelQuery>(
          latestEnvironment.current,
          channelQuery,
          {
            channelId,
          },
        ).subscribe({
          next: (data) => {
            if (data.channel)
              latestNavigation.current.navigate('Message', {
                channel: data.channel as Channel,
                users: data.channel?.memberships?.map(
                  (membership) => membership?.user as User,
                ),
              });
          },
        });
      },
    );

    return () => subscription.remove();
  }, []);

  // useEffect(() => {
  //   console.log('currentAppState', currentAppState);
  // }, [currentAppState]);

  return (
    <Stack.Navigator
      initialRouteName="MainTab"
      headerMode="screen"
      screenOptions={{
        headerBackTitle: '',
      }}>
      <Stack.Screen
        name="MainTab"
        component={TabNavigator}
        options={MainTabNavigationOptions}
      />
      <Stack.Screen
        name="ProfileUpdate"
        component={ProfileUpdate}
        options={(): StackNavigationOptions => {
          const settingButton = (): ReactElement => {
            return (
              <TouchableOpacity
                style={{
                  height: '100%',
                  minWidth: 20,
                  justifyContent: 'center',
                  marginRight: 15,
                }}
                onPress={(): void => {
                  navigation.navigate('Settings');
                }}>
                <Image style={{height: 24, width: 24}} source={IC_SETTING_W} />
              </TouchableOpacity>
            );
          };

          return {
            ...getSimpleHeader(getString('MY_PROFILE'), theme),
            headerRight: settingButton,
          };
        }}
      />
      <Stack.Screen
        name="SearchUser"
        component={SearchUser}
        options={getSimpleHeader(getString('SEARCH_USER'), theme)}
      />
      <Stack.Screen
        name="BlockedUser"
        component={BlockedUser}
        options={getSimpleHeader(getString('BLOCKED_USER'), theme)}
      />
      <Stack.Screen
        name="Report"
        component={Report}
        options={getSimpleHeader(getString('REPORT'), theme)}
      />
      <Stack.Screen
        name="Message"
        component={Message}
        options={getSimpleHeader(getString('MESSAGE'), theme)}
      />
      <Stack.Screen
        name="Settings"
        component={Settings}
        options={getSimpleHeader(getString('SETTINGS'), theme)}
      />
      <Stack.Screen
        name="ChangePw"
        component={ChangePw}
        options={getSimpleHeader(getString('PASSWORD_CHANGE'), theme)}
      />
      <Stack.Screen
        name="ChannelCreate"
        component={ChannelCreate}
        options={getSimpleHeader(getString('NEW_CHAT'), theme)}
      />
    </Stack.Navigator>
  );
}

function RootNavigator(): ReactElement {
  return (
    <View
      style={{
        flex: 1,
        flexDirection: 'column',
      }}>
      <StatusBar />
      <MainStackNavigator />
      <ProfileModal testID="modal" />
    </View>
  );
}

export default function RootNavigatorWrapper(): ReactElement {
  return (
    <ProfileModalProvider>
      <RootNavigator />
    </ProfileModalProvider>
  );
}
