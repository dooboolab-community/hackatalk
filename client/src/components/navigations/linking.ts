import {LinkingOptions} from '@react-navigation/native';
import {RootStackParamList} from './RootStackNavigator';

const linking: LinkingOptions<RootStackParamList> = {
  prefixes: ['https://hackatalk.dev', 'hackatalk://'],
  enabled: true,
  config: {
    screens: {
      MainStack: {
        screens: {
          MainTab: {
            screens: {
              Channel: 'channel',
              Friend: 'friend',
            },
          },
          ProfileUpdate: 'profile_update',
          BlockedUser: 'blocked_user',
          ChangePw: 'change_pw',
          ChannelCreate: 'channel_create',
          Message: 'message',
          PinchZoomViewPager: 'pinchzoom',
          Report: 'report',
          SearchUser: 'search_user',
          Settings: 'settings',
        },
      },
      AuthStack: {
        screens: {
          SignIn: 'sign_in',
          SignUp: 'sign_up',
          VerifyEmail: 'verify_email',
          LicenseAgreement: 'license_agreement',
          FindPw: 'find_pw',
        },
      },
    },
  },
};

export default linking;
