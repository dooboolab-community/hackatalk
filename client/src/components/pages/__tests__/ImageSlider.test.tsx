import 'react-native';

import ReactNavigation, {RouteProp} from '@react-navigation/core';
import {
  createMockNavigation,
  createTestElement,
} from '../../../../test/testUtils';

import ImageSlider from '../ImageSlider';
import React from 'react';
import {RootStackParamList} from '../../navigations/RootStackNavigator';
import {render} from '@testing-library/react-native';

const mockNavigation = createMockNavigation();

const mockRoute: RouteProp<RootStackParamList, 'ImageSlider'> = {
  key: '',
  name: 'ImageSlider',
  params: {
    images: [
      {
        uri:
          'https://images.unsplash.com/photo-1519335337423-a3357c2cd12e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2168&q=80',
      },
      {
        uri:
          'https://www.housingwire.com/wp-content/uploads/2019/09/Purple-technology-data-internet-3.jpg',
      },
      {
        uri:
          'https://images.unsplash.com/photo-1587628736664-fdc50efb57b7?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80',
      },
      {
        uri:
          'https://p.bigstockphoto.com/eIdTXLbqQilMs9xbjvcs_bigstock-Aerial-View-Of-Sandy-Beach-Wit-256330393.jpg',
      },
      {uri: 'https://avatars2.githubusercontent.com/u/7970947?v=3&s=460'},
    ],
    initialIndex: 1,
  },
};

jest.mock('@react-navigation/core', () => ({
  ...jest.requireActual<typeof ReactNavigation>('@react-navigation/core'),
  useNavigation: () => mockNavigation,
  useRoute: () => mockRoute,
}));

describe('Rendering', () => {
  it('renders without crashing', () => {
    const component = createTestElement(<ImageSlider />);
    const screen = render(component);
    const json = screen.toJSON();

    expect(json).toMatchSnapshot();
    expect(json).toBeTruthy();
  });
});
