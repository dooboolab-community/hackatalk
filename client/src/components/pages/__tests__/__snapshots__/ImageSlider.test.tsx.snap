// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering renders without crashing 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#000",
          "flexBasis": 0,
          "flexDirection": "row",
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        forwardedRef={[Function]}
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
        style={
          Object {
            "bottom": 0,
            "justifyContent": "center",
            "left": -760,
            "position": "absolute",
            "top": 0,
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": 750,
          }
        }
        testID="PINCH_ZOOM_CONTAINER"
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "https://images.unsplash.com/photo-1519335337423-a3357c2cd12e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2168&q=80",
            }
          }
          style={
            Object {
              "bottom": 0,
              "height": 1334,
              "width": 750,
            }
          }
        />
      </View>
      <View
        collapsable={false}
        forwardedRef={[Function]}
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
        style={
          Object {
            "justifyContent": "center",
            "transform": Array [
              Object {
                "translateX": 0,
              },
              Object {
                "translateY": 0,
              },
              Object {
                "scale": 1,
              },
            ],
            "width": 750,
          }
        }
        testID="PINCH_ZOOM_CONTAINER"
      >
        <View
          collapsable={false}
        >
          <Image
            resizeMode="contain"
            source={
              Object {
                "uri": "https://www.housingwire.com/wp-content/uploads/2019/09/Purple-technology-data-internet-3.jpg",
              }
            }
            style={
              Object {
                "bottom": 0,
                "height": 1334,
                "width": 750,
              }
            }
          />
        </View>
      </View>
      <View
        collapsable={false}
        forwardedRef={[Function]}
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
        style={
          Object {
            "bottom": 0,
            "justifyContent": "center",
            "left": 760,
            "position": "absolute",
            "top": 0,
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": 750,
          }
        }
        testID="PINCH_ZOOM_CONTAINER"
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "https://images.unsplash.com/photo-1587628736664-fdc50efb57b7?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
            }
          }
          style={
            Object {
              "bottom": 0,
              "height": 1334,
              "width": 750,
            }
          }
        />
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
